src/main/java/com/googleintern/wfm/ruleengine/action/DataProcessor.java

```diff
@@ -1,20 +1,69 @@
 public class DataProcessor {
 
   /** Filter out user data with invalid workgroup Id values(< 0). */
-  public static ImmutableList<UserModel> filterValidData(
+  public static ImmutableList<UserModel> filterUsersWithValidWorkgroupId(
       ImmutableList<UserModel> rawData) {
-    ImmutableList.Builder<UserModel> validDataBuilder =
-        ImmutableList.<UserModel>builder();
+    ImmutableList.Builder<UserModel> validDataBuilder = ImmutableList.<UserModel>builder();
 
     for (final UserModel data : rawData) {
       if (data.workgroupId() > 0) validDataBuilder.add(data);
     }
     return validDataBuilder.build();
   }
+
+  /**
+   * Filter out conflict user data. Conflict users are users with less role/skill ids but more
+   * assigned permissions. This function is used when we want final generated rules to assign less
+   * permissions in conflict cases.
+   *
+   * <p>Example for conflict users: User0, 1, 2 are from the same workforce and the same workgroup.
+   * User 0 has {role id = 1111, skill ids = 2222, 3333} and is assigned permissions = {AAAA, BBBB}.
+   * User 1 has {role id = 1111, skill ids = 2222} and is assigned permissions = {AAAA, BBBB, CCCC}.
+   * User 2 has {role id = 1111, skill ids = 2222} and is assigned permissions = {AAAA, CCCC}. Both
+   * User 1 and 2 are conflict users with respect to User 0. Types of Skill/role ids for User 1 and
+   * 2 are included in the types for User 0. However, User 1 and 2 have a different permission CCCC
+   * that is not included in User 0.
+   */
+  public static ImmutableList<UserModel> removeConflictUsers(ImmutableList<UserModel> rawUserData) {
+    ImmutableSet<Long> coveredConflictUsers =
+        collectAllCoveredConflictUserIds(findConflictUserIdPairs(rawUserData));
+    return rawUserData.stream()
+        .filter(user -> !coveredConflictUsers.contains(user.userId()))
+        .collect(toImmutableList());
+  }
+
+  private static ImmutableSet<Long> collectAllCoveredConflictUserIds(
+      ImmutableSetMultimap<Long, Long> conflictUserPairs) {
+    ImmutableSet.Builder<Long> coveredConflictUsers = ImmutableSet.builder();
+    conflictUserPairs
+        .keySet()
+        .forEach(key -> coveredConflictUsers.addAll(conflictUserPairs.get(key)));
+    return coveredConflictUsers.build();
+  }
+
+  private static ImmutableSetMultimap<Long, Long> findConflictUserIdPairs(
+      ImmutableList<UserModel> rawUserData) {
+    ImmutableSetMultimap.Builder<Long, Long> conflictUserPairsBuilder =
+        ImmutableSetMultimap.builder();
+    rawUserData.forEach(
+        currentUser ->
+            conflictUserPairsBuilder.putAll(
+                currentUser.userId(),
+                rawUserData.stream()
+                    .filter(comparedUser -> currentUser.isConflictUserPair(comparedUser))
+                    .map(comparedUser -> comparedUser.userId())
+                    .collect(toImmutableSet())));
+    return conflictUserPairsBuilder.build();
```

Comment:
I wonder if there is a better way to do this algorithm. Currently it is O(n^2) run time.

I can't really think of a better algorithm since you need to compare each element against the other. Something to think about, not necessarily something that needs to change

Reply:
1. mark all comparedUsers with .isConflictUserPair(comparedUser) == true to be a dirty user and currentUser  will skip if currentUser is in dirtyUserSet. 
2. mark all currentUser as visitedUser and comparedUser will skip when comparedUser is in visitedUserSet.